### PR TITLE
Add system test for Database convenience methods:

### LINKED ISSUE #3015: Add system tests for database convenience wrappers.
Add tests which exercise the higher-level `Database` convenience APIs (`read`, `execute_sql`, `run_in_transaction`, `batch`, and `snapshot`).

### PR DESCRIPTION
**Pull request by tseaver**
- `batch`
- `snapshot`
- `read`
- `run_with_transaction`
- `execute_sql`

Closes #3015.

### CODE REVIEW

File: spanner/tests/system.py [hunk 1]
```diff
@@ -228,16 +247,49 @@ def test_update_database_ddl(self):
 
         self.assertEqual(len(temp_db.ddl_statements), len(DDL_STATEMENTS))
 
+    def test_db_batch_insert_then_db_snapshot_read_and_db_read(self):
+        from google.cloud.spanner import KeySet
```
Comment (dhermes) [inline spanner/tests/system.py]: Just import globally?

File: spanner/tests/system.py [hunk 2]
```diff
@@ -228,16 +247,49 @@ def test_update_database_ddl(self):
...
+        keyset = KeySet(all_=True)
```
Comment (dhermes) [inline spanner/tests/system.py]: I noticed we make a lot of these in the system tests, could `ALL_KEYS = KeySet(all_=True)` be a global or does it require some context?

File: spanner/tests/system.py [hunk 3]
```diff
@@ -228,16 +247,49 @@ def test_update_database_ddl(self):
...
-        (1, u'Phred', u'Phlyntstone', u'phred@example.com'),
-        (2, u'Bharney', u'Rhubble', u'bharney@example.com'),
-        (3, u'Wylma', u'Phlyntstone', u'wylma@example.com'),
-    )
-    SQL = 'SELECT * FROM contacts ORDER BY contact_id'
+        retry = RetryInstanceState(_has_all_ddl)
+        retry(self._db.reload)()
+
+        with self._db.batch() as batch:
+            batch.delete(self.TABLE, keyset)
```
Comment (dhermes) [inline spanner/tests/system.py]: Why is this necessary? (i.e. is there a way to start with an empty table)

File: spanner/tests/system.py [hunk 4]
```diff
@@ -228,16 +247,49 @@ def test_update_database_ddl(self):
...
+            batch.insert(self.TABLE, self.COLUMNS, self.ROW_DATA)
+
+        with self._db.snapshot(read_timestamp=batch.committed) as snapshot:
+            from_snap = list(snapshot.read( self.TABLE, self.COLUMNS, keyset))
```
Comment (dhermes) [inline spanner/tests/system.py]: `snapshot.read( self.TABLE, ...` has an extra space after the parenthesis

File: spanner/tests/system.py [hunk 5]
```diff
@@ -228,16 +247,49 @@ def test_update_database_ddl(self):
...
+        from google.cloud.spanner import KeySet
+        keyset = KeySet(all_=True)
+
+        retry = RetryInstanceState(_has_all_ddl)
+        retry(self._db.reload)()
+
+        with self._db.batch() as batch:
+            batch.delete(self.TABLE, keyset)
+
+        def _unit_of_work(transaction, test):
```
Comment (dhermes) [inline spanner/tests/system.py]: Can you make this a `@staticmethod` on the `TestCase`?

File: spanner/tests/system.py (hunk 1 above)
Comment (tseaver) [inline spanner/tests/system.py]: OK,

File: spanner/tests/system.py (hunk 2 above)
Comment (tseaver) [inline spanner/tests/system.py]: I'll make it a class attribute of the base class which has the other constants used to set up the tables.

File: spanner/tests/system.py (hunk 3 above)
Comment (tseaver) [inline spanner/tests/system.py]: Races about previous uses of the same table (we can't create new tables cheaply on the fly).

File: spanner/tests/system.py (hunk 5 above)
Comment (tseaver) [inline spanner/tests/system.py]: Why?  The "normal" use of  `run_in_transaction` is to pass in a function with additional varargs / kwargs as needed, which is exactly what we are doing here.

**Comment by tseaver** [comment]
Should I [follow suit](https://github.com/GoogleCloudPlatform/google-cloud-python/pull/3384#issuecomment-300004596) and push the updated branch to the origin to get a build with system tests?  We could close this PR and re-open there.

**Comment by dhermes** [comment]
@tseaver Yes I think you should push this branch to the GCP origin to get system tests running. You don't need to close this PR though, the tests will "just work" (i.e. be associated here).

File: spanner/tests/system.py (hunk 5 above)
Comment (dhermes) [inline spanner/tests/system.py]: More [dynamically created function allergy][1]

[1]: https://github.com/GoogleCloudPlatform/google-cloud-python/pull/3384#discussion_r115375627

**Comment by tseaver** [comment]
Pushed to `upstream`.

File: spanner/tests/system.py (hunk 5 above)
Comment (tseaver) [inline spanner/tests/system.py]: *De gustibus*

**Comment by tseaver** [comment]
@dhermes New system tests pass on CI.  Any further issues?

**Comment by dhermes** [comment]
LGTM
